### PR TITLE
Additional tracing and retries on compiler errors

### PR DESCRIPTION
**Pull request by fabiocav**
This enables simple retries on compiler failures (actual compiler failures as opposed to compilation errors), but more importantly, it traces some important information we're currently missing today.

### CODE REVIEW

File: src/WebJobs.Script/Description/DotNet/Compilation/CSharp/CSharpCompilation.cs [hunk 1]
```diff
@@ -91,29 +91,36 @@ private static bool IsOrUsesAssemblyType(ITypeSymbol typeSymbol, IAssemblySymbol
 
         public Assembly Emit(CancellationToken cancellationToken)
         {
-            using (var assemblyStream = new MemoryStream())
+            try
             {
-                using (var pdbStream = new MemoryStream())
+                using (var assemblyStream = new MemoryStream())
```
Comment (mamaso) [inline src/WebJobs.Script/Description/DotNet/Compilation/CSharp/CSharpCompilation.cs]: nit: combine these usings to remove nesting

**Comment by fabiocav** [comment]
Resetting it does not trigger compilation. This will allow subsequent invocations to trigger compilation again, as opposed to the current behavior where it returns the cached exception.

The exception is just re-thrown by the invoker. It allows it to bubble up, which will be handled as an invocation failure when in the invocation path or function initialization error on initial loading.

We can decide to be more aggressive down the road, but this allows for lazy recovery and gives us more information on compiler level errors.

**Comment by fabiocav** [comment]
@brettsam @mathewc, do you guys want to take a quick look at this?

File: src/WebJobs.Script/Description/Compilation/CompilerErrorException.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+using System.Linq;
+using System.Runtime.Serialization;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Scripting;
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    [Serializable]
+    public sealed class CompilerErrorException : Exception
```
Comment (brettsam) [inline src/WebJobs.Script/Description/Compilation/CompilerErrorException.cs]: nit: `CompilerErrorException` and `CompilationErrorException` sound very similar -- is there another name we can give this? Even `CompilerException` sounds different enough to me...

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs [hunk 3]
```diff
@@ -189,6 +191,20 @@ public override void OnError(Exception ex)
...
+                TraceOnPrimaryHost("Script compiler error", TraceLevel.Error, _compilationService.GetType().Name, exc);
+
+                // Compiler errors are often sporadic, so we'll attempt to reset the loader here to avoid
+                // caching the compiler error and leaving the function hopelessly broken
+                if (++_compilerErrorCount < 3)
+                {
+                    _functionLoader.Reset();
+
+                    TraceOnPrimaryHost("Function loader reset. Failed compilation result will not be cached.", TraceLevel.Info);
+                }
```
Comment (brettsam) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: These messages are mostly for us, aren't they? Is there anything actionable for the user?

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs [hunk 4]
```diff
@@ -189,6 +191,20 @@ public override void OnError(Exception ex)
                 TraceCompilationDiagnostics(exc.Diagnostics, LogTargets.User);
                 throw;
             }
+            catch (CompilerErrorException exc)
+            {
+                TraceOnPrimaryHost("Script compiler error", TraceLevel.Error, _compilationService.GetType().Name, exc);
```
Comment (brettsam) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: On first glance, this sounds like I've got an issue with my code compiling -- can we make it clear that that's not the case?

Comment (brettsam) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: If these are sporadic, is it possible for one host to have an exception while the others don't? If that's the case, we shouldn't just be tracing from the primary host here, should we?

Comment (fabiocav) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: Sure... do you have any suggestions? Would removing script and saying `Compiler error` only help? Perhaps something like `Compiler service error`?

Comment (fabiocav) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: Yeah, that's a good point. I'll update this.

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs (hunk 3 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: Those are tracing messages to give us better visibility when those problems happen. We can flag those as system only.

File: src/WebJobs.Script/Description/Compilation/CompilerErrorException.cs (hunk 2 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/Compilation/CompilerErrorException.cs]: Yeah, they sound similar. I'll change this to `CompilationServiceException`

**Comment by fabiocav** [comment]
Addressed all of the feedback. Can you guys take a quick look? Trying to close all of this up today.

**Pull request merged**